# nf-core/methylong: Citations

## [nf-core](https://pubmed.ncbi.nlm.nih.gov/32055031/)

> Ewels PA, Peltzer A, Fillinger S, Patel H, Alneberg J, Wilm A, Garcia MU, Di Tommaso P, Nahnsen S. The nf-core framework for community-curated bioinformatics pipelines. Nat Biotechnol. 2020 Mar;38(3):276-278. doi: 10.1038/s41587-020-0439-x. PubMed PMID: 32055031.

## [Nextflow](https://pubmed.ncbi.nlm.nih.gov/28398311/)

> Di Tommaso P, Chatzou M, Floden EW, Barja PP, Palumbo E, Notredame C. Nextflow enables reproducible computational workflows. Nat Biotechnol. 2017 Apr 11;35(4):316-319. doi: 10.1038/nbt.3820. PubMed PMID: 28398311.

## Pipeline tools

- [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)

> Andrews, S. (2010). FastQC: A Quality Control Tool for High Throughput Sequence Data [Online].

- [MultiQC](https://pubmed.ncbi.nlm.nih.gov/27312411/)

> Ewels P, Magnusson M, Lundin S, Käller M. MultiQC: summarize analysis results for multiple tools and samples in a single report. Bioinformatics. 2016 Oct 1;32(19):3047-8. doi: 10.1093/bioinformatics/btw354. Epub 2016 Jun 16. PubMed PMID: 27312411; PubMed Central PMCID: PMC5039924.

- [samtools](https://github.com/samtools/samtools)

> Petr Danecek, James K Bonfield, Jennifer Liddle, John Marshall, Valeriu Ohan, Martin O Pollard, Andrew Whitwham, Thomas Keane, Shane A McCarthy, Robert M Davies, Heng Li (2021) Twelve years of SAMtools and BCFtools. GigaScience, Volume 10, Issue 2, February 2021, giab008, https://doi.org/10.1093/gigascience/giab008

- [gunzip](https://www.gnu.org/software/gzip/manual/gzip.html)

- [pigz](https://zlib.net/pigz/)

- [minimap2](https://github.com/lh3/minimap2)

> Li, H. (2018). Minimap2: pairwise alignment for nucleotide sequences. Bioinformatics, 34:3094-3100. doi:10.1093/bioinformatics/bty191

> Li, H. (2021). New strategies to improve minimap2 alignment accuracy. Bioinformatics, 37:4572-4574. doi:10.1093/bioinformatics/btab705

- [pbmm2](https://github.com/PacificBiosciences/pbmm2)

- [dorado](https://github.com/nanoporetech/dorado)

> licence under: Oxford Nanopore Technologies Plc.

- [porechop](https://github.com/rrwick/Porechop)

> Wick RR, Judd LM, Gorrie CL, Holt KE. Completing bacterial genome assemblies with multiplex MinION sequencing. Microb Genom. 2017;3(10):e000132. Published 2017 Sep 14. doi:10.1099/mgen.0.000132

- [modkit](https://github.com/nanoporetech/modkit)

> licence under: Oxford Nanopore Technologies Plc.

- [pb-CpG-tools](https://github.com/PacificBiosciences/pb-CpG-tools)

> licence under: Pacific Biosciences Software License Agreement

- [Clair3](https://github.com/HKU-BAL/Clair3)

- [WhatsHap](https://github.com/whatshap/whatshap)

> Marcel Martin, Murray Patterson, Shilpa Garg, Sarah O. Fischer, Nadia Pisanti, Gunnar W. Klau, Alexander Schoenhuth, Tobias Marschall. _WhatsHap: fast and accurate read-based phasing_. bioRxiv 085050; doi: https://doi.org/10.1101/085050

- [gawk](https://www.gnu.org/software/gawk)
- [DSS](https://bioconductor.org/packages/release/bioc/html/DSS.html)

> Wu H, Wang C, Wu Z (2013). A new shrinkage estimator for dispersion improves differential expression detection in RNA-seq data. _Biostatistics_. [doi:10.1093/biostatistics/kxs033](https://doi.org/10.1093/biostatistics/kxs033).
>
> Feng H, Conneely K, Wu H (2014). A bayesian hierarchical model to detect differentially methylated loci from single nucleotide resolution sequencing data. _Nucleic acids research_. [doi:10.1093/nar/gku154](https://doi.org/10.1093/nar/gku154).
>
> Wu H, Xu T, Feng H, Chen L, Li B, Yao B, Qin Z, Jin P, Conneely K (2015). Detection of differentially methylated regions from whole-genome bisulfite sequencing data without replicates. _Nucleic acids research_. [doi: 10.1093/nar/gkv715](https://doi.org/ 10.1093/nar/gkv715).
>
> Park Y, Wu H (2016). Differential methylation analysis for BS-seq data under general experimental design. _Bioinformatics_. [doi:10.1093/bioinformatics/btw026](https://doi.org/10.1093/bioinformatics/btw026).

- [jasmine](https://github.com/pacificbiosciences/jasmine/)
- [ccsmeth](https://github.com/PengNi/ccsmeth)

> Ni P, Nie F, Zhong Z. _et al._ DNA 5-methylcytosine detection and methylation phasing using PacBio circular consensus sequencing. _Nat Commun_ **14**, 4054 (2023). https://doi.org/10.1038/s41467-023-39784-9

- [fibertools](https://github.com/fiberseq/fibertools-rs)

> Jha A, Bohaczuk SC, Mao Y, Ranchalis J, Mallory BJ, Min AT, Hamm MO, Swanson E, Dubocanin D, Finkbeiner C, Li T, Whittington D, Noble WS, Stergachis AB, & Vollger MR (2024). DNA-m6A calling and integrated long-read epigenetic and genetic analysis with fibertools. Genome Research. https://doi.org/10.1101/gr.279095.124

## Software packaging/containerisation tools

- [Anaconda](https://anaconda.com)

  > Anaconda Software Distribution. Computer software. Vers. 2-2.4.0. Anaconda, Nov. 2016. Web.

- [Bioconda](https://pubmed.ncbi.nlm.nih.gov/29967506/)

  > Grüning B, Dale R, Sjödin A, Chapman BA, Rowe J, Tomkins-Tinch CH, Valieris R, Köster J; Bioconda Team. Bioconda: sustainable and comprehensive software distribution for the life sciences. Nat Methods. 2018 Jul;15(7):475-476. doi: 10.1038/s41592-018-0046-7. PubMed PMID: 29967506.

- [BioContainers](https://pubmed.ncbi.nlm.nih.gov/28379341/)

  > da Veiga Leprevost F, Grüning B, Aflitos SA, Röst HL, Uszkoreit J, Barsnes H, Vaudel M, Moreno P, Gatto L, Weber J, Bai M, Jimenez RC, Sachsenberg T, Pfeuffer J, Alvarez RV, Griss J, Nesvizhskii AI, Perez-Riverol Y. BioContainers: an open-source and community-driven framework for software standardization. Bioinformatics. 2017 Aug 15;33(16):2580-2582. doi: 10.1093/bioinformatics/btx192. PubMed PMID: 28379341; PubMed Central PMCID: PMC5870671.

- [Docker](https://dl.acm.org/doi/10.5555/2600239.2600241)

  > Merkel, D. (2014). Docker: lightweight linux containers for consistent development and deployment. Linux Journal, 2014(239), 2. doi: 10.5555/2600239.2600241.

- [Singularity](https://pubmed.ncbi.nlm.nih.gov/28494014/)

  > Kurtzer GM, Sochat V, Bauer MW. Singularity: Scientific containers for mobility of compute. PLoS One. 2017 May 11;12(5):e0177459. doi: 10.1371/journal.pone.0177459. eCollection 2017. PubMed PMID: 28494014; PubMed Central PMCID: PMC5426675.

- [charliecloud](https://hpc.github.io/charliecloud/)

  > Reid Priedhorsky and Tim Randles. “Charliecloud: Unprivileged containers for user-defined software stacks in HPC”, 2017. In Proc. Supercomputing. DOI: 10.1145/3126908.3126925.
